build_web_compilers/lib/src/workers.dart

```diff
@@ -83,4 +90,125 @@ final dartdevcDriverResource = new Resource<BazelWorkerDriver>(
 
 final sdkDir = cli_util.getSdkPath();
 
-final dart2jsPath = p.join(sdkDir, 'bin', 'dart2js$_scriptExtension');
+/// Manages a shared set of persistent dart2js workers.
+Dart2JsBatchWorker get _dart2jsWorker {
+  _dart2jsWorkersAreDoneCompleter ??= new Completer<Null>();
+  return __dart2jsWorker ??= new Dart2JsBatchWorker(() => Process.start(
+      p.join(sdkDir, 'bin', 'dart2js$_scriptExtension'), ['--batch'],
+      workingDirectory: scratchSpace.tempDir.path));
+}
+
+Dart2JsBatchWorker __dart2jsWorker;
+
+/// Resource for fetching the current [Dart2JsBatchWorker] for dart2js.
+final dart2JsWorkerResource = new Resource<Dart2JsBatchWorker>(
+    () => _dart2jsWorker, beforeExit: () async {
+  await _dart2jsWorker.terminateWorkers();
+  _dart2jsWorkersAreDoneCompleter.complete();
+  _dart2jsWorkersAreDoneCompleter = null;
+});
+
+/// Manages a persistent dart2js worker running in batch mode and schedules jobs
+/// one at a time.
+class Dart2JsBatchWorker {
+  final Future<Process> Function() _spawnWorker;
+
+  Stream<String> __workerStderrLines;
+  Stream<String> get _workerStderrLines {
+    assert(__worker != null);
+    return __workerStderrLines ??= __worker.stderr
+        .transform(UTF8.decoder)
+        .transform(const LineSplitter())
+        .asBroadcastStream();
+  }
+
+  Stream<String> __workerStdoutLines;
+  Stream<String> get _workerStdoutLines {
+    assert(__worker != null);
+    return __workerStdoutLines ??= __worker.stdout
+        .transform(UTF8.decoder)
+        .transform(const LineSplitter())
+        .asBroadcastStream();
+  }
+
+  Process __worker;
+  Future<Process> get _worker async {
+    __worker ??= await _spawnWorker();
+    return __worker;
+  }
+
+  final _workQueue = new Queue<_Dart2JsJob>();
+
+  bool _queueIsActive = false;
+
+  Dart2JsBatchWorker(this._spawnWorker);
+
+  Future<Dart2JsResult> compile(List<String> args) async {
+    var job = new _Dart2JsJob(args);
+    _workQueue.add(job);
+    if (!_queueIsActive) _startWorkQueue();
+    return job.result;
+  }
+
+  void _startWorkQueue() {
```

Comment:
[[optional] Is there any duplication we can factor out?

Reply:
not really... other than adding the lines to the buffer but the logic is so trivial its not worth breaking out imo